byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+	}
+
+	sc = append(sc, NewStateChange(Create, inst.DeriveID(""),
+		ContractDeferredID, dataBuf, darcID))
+	return
+}
+
+func (c *contractDeferred) Invoke(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	// This method should do the following:
+	//   - Handle the "addProof" invocation
+	//   - Handle the "execProposedTx" invocation
+	//
+	// Invoke:addProof should have the following input argument:
+	//   - identity darc.Identity
+	//   - signature []byte
+	//   - index uint32 (index of the instruction wrt the transaction)
+	err2 := c.checkInvoke(rst, inst.Invoke)
+	if err2 != nil {
+		return nil, nil, errors.New("checks of invoke failed: " + err2.Error())
+	}
+
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	switch inst.Invoke.Command {
+	case "addProof":
+		// This invocation appends the identity and the corresponding signature,
+		// which is based on the stored instruction hash (in instructionHashes)
+
+		// Get the given index
+		indexBuf := inst.Invoke.Args.Search("index")
+		if indexBuf == nil {
+			return nil, nil, errors.New("index args is nil")
+		}
+		index := binary.LittleEndian.Uint32(indexBuf)
+
+		// Check if the index is in range
+		numInstruction := len(c.DeferredData.ProposedTransaction.Instructions)
+		if index >= uint32(numInstruction) {
+			return nil, nil, fmt.Errorf("index is out of range (%d >= %d)", index, numInstruction)
+		}
+
+		// Get the given Identity
+		identityBuf := inst.Invoke.Args.Search("identity")
+		if identityBuf == nil {
+			return nil, nil, errors.New("identity args is nil")
+		}
+		identity := darc.Identity{}
+		err = protobuf.Decode(identityBuf, &identity)
+		if err != nil {
+			return nil, nil, errors.New("couldn't decode Identity")
+		}
+
+		// Get the given signature
+		signature := inst.Invoke.Args.Search("signature")
+		if signature == nil {
+			return nil, nil, errors.New("signature args is nil")
+		}
+		// Update the contract's data with the given signature and identity
+		c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities = append(c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities, identity)
+		c.DeferredData.ProposedTransaction.Instructions[index].Signatures = append(c.DeferredData.ProposedTransaction.Instructions[index].Signatures, signature)
+		// Save and send the modifications
+		cosiDataBuf, err2 := protobuf.Encode(&c.DeferredData)
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't encode DeferredData")
+		}
+		sc = append(sc, NewStateChange(Update, inst.InstanceID,
+			ContractDeferredID, cosiDataBuf, darcID))
+		return
+	case "execProposedTx":
+		// This invocation tries to execute the transaction stored with the
+		// "Spawn" invocation. If it is successful, this invocation fills the
+		// "ExecResult" field of the "deferredData" struct.
```

Comment:
could you add a comment here explaining why something like `processOneTx` cannot be usef?